src/System.Memory/src/System/Span.cs

```diff
@@ -323,36 +322,82 @@ public void CopyTo(Span<T> destination)
                 return false;
 
-            // TODO: This is a tide-over implementation as we plan to add a overlap-safe cpblk-based api to Unsafe. (https://github.com/dotnet/corefx/issues/13427)
             unsafe
             {
                 ref T src = ref DangerousGetPinnableReference();
                 ref T dst = ref destination.DangerousGetPinnableReference();
                 IntPtr srcMinusDst = Unsafe.ByteOffset<T>(ref dst, ref src);
-                int length = _length;
+                IntPtr dstMinusSrc = Unsafe.ByteOffset<T>(ref src, ref dst);
 
                 bool srcGreaterThanDst = (sizeof(IntPtr) == sizeof(int)) ? srcMinusDst.ToInt32() >= 0 : srcMinusDst.ToInt64() >= 0;
                 if (srcGreaterThanDst)
                 {
-                    // Source address greater than or equal to destination address. Can do normal copy.
-                    for (int i = 0; i < length; i++)
+                    IntPtr tailDiff = Unsafe.ByteOffset<T>(ref Unsafe.Add<T>(ref dst, destLength), ref src);
+                    bool isOverlapped = (sizeof(IntPtr) == sizeof(int)) ? tailDiff.ToInt32() < 0 : tailDiff.ToInt64() < 0;
+                    if (!isOverlapped && !SpanHelpers.IsReferenceOrContainsReferences<T>())
+                    {
+                        CopyBlocks(ref dst, ref src, length);
+                    }
+                    else
                     {
-                        Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        // Source address greater than or equal to destination address. Can do normal copy.
+                        for (int i = 0; i < length; i++)
+                        {
+                            Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        }
                     }
+
+                    return true;
                 }
-                else
+
+                bool dstGreaterThanSrc = (sizeof(IntPtr) == sizeof(int)) ? dstMinusSrc.ToInt32() >= 0 : dstMinusSrc.ToInt64() >= 0;
+                if (dstGreaterThanSrc)
                 {
-                    // Source address less than destination address. Must do backward copy.
-                    int i = length;
-                    while (i-- != 0)
+                    IntPtr tailDiff = Unsafe.ByteOffset<T>(ref Unsafe.Add<T>(ref src, length), ref dst);
+                    bool isOverlapped = (sizeof(IntPtr) == sizeof(int)) ? tailDiff.ToInt32() < 0 : tailDiff.ToInt64() < 0;
+                    if (!isOverlapped && !SpanHelpers.IsReferenceOrContainsReferences<T>())
+                    {
+                        CopyBlocks(ref dst, ref src, length);
+                    }
+                    else
                     {
-                        Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        // Source address less than destination address. Must do backward copy.
+                        int i = length;
+                        while (i-- != 0)
```

Comment:
I never changed that part. Not sure if there is a performance reason it was done.